### PR TITLE
feat(auth): authority profile - production / principal toggle

### PR DESCRIPTION
**Pull request by bbasketballer75**
## feat(auth): authority profile — production / principal toggle

Adds an `authority_profile` config key (default `production`, opt-in `principal`) and a single source-of-truth rule module (`hermes_cli/authority`) for authority decisions, replacing rules duplicated between SOUL.md and the runtime guard.

**Enforcement added by this PR: one boundary.** The only action gated in production code today is the `config.yaml` write (`tools/file_tools_write_guards.py` -> `authority.allows("edit_config_yaml")`, wired through `tools/file_tools.py`). The other declared actions (`install_dependency`, `create_skill`, `install_plugin`) exist in the module and are unit-tested, but have **no production callers yet** — `tools/skills_hub_install.py` and `tools/plugins.py` contain no authority reference. See *Scope* at the end of this description for the full enforced/not-enforced split.

### Files
- `hermes_cli/authority.py` (new, 323 lines): the rule module — `AuthorityProfile`, `allows(action, target=...)`, `log_action(...)`, profile resolution, env override (`HERMES_AUTHORITY_PROFILE`), audit log to `~/.hermes/state/authority-audit.jsonl`.
- `tests/hermes_cli/test_authority.py` (new, 450 lines): unit tests covering default profile, principal override, env override, each guarded action (file edit / dep install / skill create / plugin install), audit-log shape, and import safety.
- `tools/file_tools_write_guards.py` (22-line edit): `_check_sensitive_path` consults `authority.allows("edit_config_yaml")` before the hard-deny on `config.yaml`. Production profile keeps the existing deny; principal profile + audit log downgrades it to allow. Absolute denies (sensitive system paths, approval globs, hermes-agent upstream, cookie/secret files) are unchanged.

### Upstream check (still needed)
Verified on `upstream/main` at the SHA used for the branch base:
- `git cat-file -e upstream/main:hermes_cli/authority.py` → exit 128 (path does not exist).
- `git ls-tree upstream/main -- hermes_cli/ | grep -i authority` → no match.
- `git show upstream/main:tools/file_tools_write_guards.py | grep -i -E "authority|principal"` → no `authority`/`principal` references (only the unrelated word "production" in a docstring). The `_check_sensitive_path` body hard-denies `config.yaml` unconditionally.

So no equivalent exists upstream; this change is not redundant.

### Port note
`git cherry-pick 6900a6edcc` applied cleanly on top of `upstream/main` (no conflicts). The 22-line guard edit is byte-identical to the original local commit; the two new files were added verbatim. No manual porting was required.

### Tests
- New suite `tests/hermes_cli/test_authority.py`: **32/32 pass** (Python 3.11.15, pytest 9.1.1).
- Same guard suite run against unmodified `upstream/main` as baseline:
  - baseline (upstream/main): 21 failed, 143 passed, 2 skipped
  - this branch: 21 failed, 207 passed (+32 from new file), 2 skipped
  - failing-test-name diff: **identical set** (0 introduced failures). The 21 pre-existing failures are unrelated Windows-only issues (POSIX `/etc/`, `/boot/`, `/proc/` assertions; atomic-write/BOM patch-handler regressions) — none touch `hermes_cli/authority` or the modified guard path.

Test command: `pytest tests/hermes_cli/test_authority.py tests/tools/test_binary_document_write_guard.py tests/tools/test_file_read_guards.py tests/tools/test_file_tools.py tests/tools/test_file_write_safety.py`.

---

## Scope — what this does and does not enforce

Corrected after review, because the original description implied a complete authority boundary.

**Enforced:**

- `config.yaml` writes: hard-denied under `production`; allowed under `principal` with an audit entry.
- **The policy keys themselves are never agent-writable.** `authority_profile` and `approvals` are
  compared against the current file on every whole-file write and refused if changed, under either
  profile; fragment (`patch`) edits to `config.yaml` are refused under `principal` because they cannot
  be inspected. Without this, `principal` was self-granting and `approvals` was switchable from inside
  the agent.
- `authority_profile` is a registered key in `DEFAULT_CONFIG`, so `hermes config set` knows it.

**NOT enforced — policy data only:**

- The other labels in `PRINCIPAL_ALLOWS` (`edit_soul_md`, `edit_env`, `plugin_install`, `skill_create`,
  package installs, …) have **no production call sites**. They describe intent; nothing consults them
  yet. Wiring each mutation boundary is a design decision this PR deliberately leaves open.
- The profile, config path and cache are resolved at import, so a multiplexed process could in
  principle let one profile's mode colour another's. Resolving per context-local profile identity is
  a separate change.

Say the word if you would rather this be narrowed further (module + config key only) or expanded to
wire the mutation sites; both are reasonable and I would rather follow your direction than guess.

## Decision on scope (this PR's final position)

After the review discussion, this PR deliberately stays a **config-guard policy layer** and does not
grow `allows()` calls at the other mutation sites. Reasons, in order of weight:

1. Where the authority boundary belongs in Hermes is a maintainer decision. Wiring skill-create,
   plugin-install and package-install paths from a contributor PR would be inventing product
   direction — the same failure mode as shipping a well-tested helper for a concept the project does
   not have.
2. What *is* enforced here is small, testable and safe: the `config.yaml` guard, plus policy-key
   immutability so the grant cannot be self-modified. All 13 tests target that.
3. The process-global resolution is left as-is on purpose. Resolving per context-local identity would
   change what the live guard reads, and the current behaviour fails in the *safe* direction (process
   home = the machine-root config, which is where the profile is meant to be set). Changing it for a
   multiplex case this install does not use is not worth the risk to the working gate.

The unenforced labels in `PRINCIPAL_ALLOWS` are therefore documentation of intent, not enforcement,
and the description says so. If you would rather have the mutation-site wiring, I am glad to build it
to your design — it just should not be guessed at.

### CODE REVIEW

**Review by gaoanze888 — COMMENTED** [review]
The parser defaults invalid/missing values to production and focused authority tests pass, but the feature is not yet a safe or complete authority boundary at exact head `12cdcbd968348a76e95c4505384c28b146cca9ec`:

1. State is process-global: `HERMES_HOME`, config path, and cache are fixed at import. In multiplex contexts, whichever profile loads first can leak production/principal mode to the other. Resolve/cache by context-local profile identity and test interleaving.
2. Principal mode allows `edit_config_yaml`, which bypasses the protected-config hard refusal. The agent can then disable approvals or alter protected instruction/security settings—including its own authority source. Absolute-deny labels are not wired into approval enforcement. Do not grant self-modification of the authorization policy.
3. Most declared capabilities have no production callers; terminal/package/plugin/skill/protected-file paths do not consult `allows()`. The description therefore overstates enforcement.
4. `authority_profile` is absent from defaults/schema/CLI validation, so no auditable UI/API/CLI toggle parity exists.

Define a profile-scoped, immutable-per-operation authority decision, wire every claimed boundary, add a separately authenticated transition API, and test privilege-escalation attempts. If audit is a security guarantee, record outcomes and define fail-closed behavior.

**Comment by bbasketballer75** [comment]
Thanks both — two of these were real, and the second was a genuine escalation hole rather than a
style issue.

**1. `authority_profile` was not a known config key — fixed.** You are right that the advertised
`hermes config set authority_profile principal` path warned about an unknown key and that the default
existed only as a literal inside the module. It is now registered in `hermes_cli/config_defaults.py`
(`DEFAULT_CONFIG`) with both accepted values and what they mean, so there is one canonical definition.

**2. Principal could rewrite the keys that define principal — fixed.** This is the one I want to name
plainly: `_check_sensitive_path` converted the **whole** `config.yaml` hard-deny into an allow under
principal, so the agent could flip `authority_profile: production` → `principal` (self-escalation) and
could turn off `approvals` — the exact exec-approval gate the comment around that code says this guard
family exists to protect. Adding `edit_config_yaml` to `ABSOLUTE_DENIES` would not have helped, because
that set is data, not an enforcement point.

Fixed by narrowing the grant instead of removing it, so principal keeps the capability it exists for:

- `_check_authority_policy_write` compares `authority_profile` and `approvals` between the current file
  and the proposed content and refuses a change under **either** profile. Fails closed — YAML that will
  not parse as a mapping counts as a change.
- `_check_authority_policy_patch` refuses *fragment* edits to `config.yaml` under principal, since
  `patch` supplies only old/new fragments and cannot be shown to leave the policy keys alone. A
  whole-file `write_file` is still allowed and **is** inspected.
- Every other key stays editable under principal, including a whole-file rewrite that carries the
  unchanged policy blocks — the common case, and refusing it would make the profile useless.

**3. "Most declared capabilities have no production callers" — accepted; description narrowed.**
Correct, and the PR description now says so instead of implying a complete boundary. `allows()` is
consulted at the `config.yaml` guard and nowhere else; the other labels in `PRINCIPAL_ALLOWS` are
policy data today, not enforcement. Wiring each mutation site (skill create, plugin install, dependency
paths) is a larger change and a decision about where the boundary belongs, which I would rather take
from a maintainer than invent here.

**4. Process-global state — accepted, not fixed in this PR.** You are right that `HERMES_HOME`, the
config path and the cache are resolved at import, so in a multiplexed process whichever profile loads
first can colour the others. Fixing it properly means resolving and caching per context-local profile
identity, with interleaving tests. That is a design change I am not folding into this PR silently —
it is listed as a known limitation instead.

Verification: 13 new tests in `tests/hermes_cli/test_authority_policy_guard.py` cover the grant being
intact, both escalation directions, disabling approvals, patch refusal, production being untouched, and
unknown authority state denying rather than crashing. Guards + authority suites: **99 passed, 7 failed**
— the failing set is byte-identical to unmodified `main` on this host (7 pre-existing macOS-path
failures), so nothing new broke.

**Review by gaoanze888 — COMMENTED** [review]
Rechecked exact head `0c1d53fb1c4b6b8ed9992b79943285fe6c5edf6b`; the new focused authority suites pass 45/45. Registering the config key and making malformed/changed whole-file policy writes fail closed addresses the direct self-modification case, but the guard is still bypassable through path replacement:

`write_file_tool()` performs `_check_sensitive_path()` and `_check_authority_policy_write()` before resolving/locking the eventual target. Later it independently resolves the path and writes that result. If an alias points to a benign file during the guards and is swapped to a symlink to `config.yaml` before the later resolution, the operation writes the protected config without policy comparison. The same separation allows a concurrent legitimate policy edit after comparison to be silently overwritten by the previously approved whole-file content.

Please resolve one canonical target once, lock that identity, then run the sensitive-path/policy comparison and write against that same frozen target; validate identity/content again at commit if the backend cannot provide a bound handle. Add a barrier test that swaps a benign alias to the config symlink between guard and write, plus a concurrent policy-update test. Fragment patch has the analogous precheck→re-resolve gap and should share the same bound-target mechanism.

The broader process-global profile/cache and unwired-capability limitations remain as documented/narrowed rather than fixed by this increment.

File: hermes_cli/authority.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        _cache_profile = _read_profile_from_config()
+        _cache_ts = now
+    return _cache_profile
+
+
+def allows(action: str) -> bool:
+    """Return True if ``action`` is permitted under the active profile.
+
+    Order of checks:
+      1. Absolute denies win over everything (always fail-closed).
```
[the remark this answers is not in the document]
Comment (bbasketballer75) [reply hermes_cli/authority.py:149]: You're right, and I corrected the description rather than leaving it overstating the change.

Verified at head `0c1d53fb1c`: the only **production** caller is the `config.yaml` write path — `tools/file_tools_write_guards.py` calls `authority.allows("edit_config_yaml")`, wired through `tools/file_tools.py`. `tools/skills_hub_install.py` and `tools/plugins.py` contain **no authority reference at all**.

The opening paragraph claimed the key "gates file edits, dependency installs, skill creation, and plugin installation". That was wrong; it now reads:

> **Enforcement added by this PR: one boundary.** The only action gated in production code today is the `config.yaml` write … The other declared actions (`install_dependency`, `create_skill`, `install_plugin`) exist in the module and are unit-tested, but have **no production callers yet** — `tools/skills_hub_install.py` and `tools/plugins.py` contain no authority reference.

pointing at the enforced/not-enforced split already in the Scope section. The module contract is unchanged; the description no longer claims enforcement that isn't wired.

File: hermes_cli/authority.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
+"""Authority profile — controls what the agent can do per session.
+
+Two profiles, opt-in by Austin via ``config.yaml``::
+
+    authority_profile: production   # default — current safety floors
+    authority_profile: principal    # lifts edit/create/install rules
```
[the remark this answers is not in the document]
Comment (bbasketballer75) [reply hermes_cli/authority.py:6]: Fixed on the current head. `hermes_cli/config_defaults.py:47` contains `"authority_profile": "production"`, with the explanatory block at `:43`, and `config_defaults.py` is one of this PR's changed files. So `hermes config set authority_profile principal` no longer lands as an unknown key, and the advertised option has one canonical definition.

Correct when raised — the fix is in this same PR.

**Comment by bbasketballer75** [comment]
@gaoanze888 — I see the path-replacement bypass finding and it is a real escalation hole: the `_check_sensitive_path` / `_check_authority_policy_write` guards run before the eventual target is resolved and locked, so an alias swap between guard and write (or a concurrent legitimate policy edit between comparison and write) can route around the policy comparison. The bound-target approach you describe — resolve one canonical target, lock that identity, run guards and write against the frozen handle, revalidate identity at commit — is the right shape, and the barrier test (benign alias → config symlink between guard and write) is the test I owe you. I am not going to push a fix without working through the bound-handle contract end-to-end first; that needs a real design conversation, not a tactical patch on top of this PR. Tracking the bound-target mechanism and the concurrent policy-update test as a follow-up rather than smuggling them into a stale-PR correction; happy to take the design direction from you on it.

**Comment by FriendlyCityJohnstown** [comment]
## Status nudge

This PR is open and awaiting review. There is maintainer signal on it that I want to surface so we don't lose context:

> @gaoanze888 (review COMMENTED): The parser defaults invalid/missing values to production and focused authority tests pass, but the feature is not yet a safe or complete authority boundary at exact head `12cdcbd968348a76e95c4505384c28b146cca9ec`:

1. State is process-global: `HERM

I have not pushed anything new to the branch since the prior comment. If anything above is no longer accurate against current `main`, please let me know which finding to address first; I will rebase and push a follow-up before doing anything else on this PR.

**Comment by bbasketballer75** [comment]
Rebased onto current `main` (`a25cf4d77d`).

**Branch tip:** `0c1d53fb1c4b` → `4a90051bab95`

## Rebase notes

One conflict, in `tools/file_tools_write_guards.py`, and it was confined to a **docstring**:

- `main` added an NT/device-namespace guard (`get_nt_namespace_error(filepath, verb="Write")`) at the top of `_check_sensitive_path`.
- This branch rewrote that function's docstring to document the authority-profile behaviour.

Both were kept — `main`'s guard runs first, then the authority-profile logic proceeds unchanged. No behavioural change to either side.

## Verification

```
$ pytest tests/hermes_cli/test_authority.py tests/hermes_cli/test_authority_policy_guard.py -q
45 passed
```

Regression check on the module this PR modifies:

```
$ pytest tests/tools/test_file_write_safety.py tests/tools/test_file_tools.py -q
10 failed, 94 passed
```

Pristine `main`, same two files, same machine:

```
$ pytest tests/tools/test_file_write_safety.py tests/tools/test_file_tools.py -q
10 failed, 94 passed
```

**The same 10 failures occur on pristine `main`** (Windows-specific path-check expectations), so the rebase introduces no regressions.

Diff is unchanged from the original: 6 files, +1052/−2, including the new `hermes_cli/authority.py`.
